src/memo_ingredient_indices.rs

```diff
@@ -49,66 +29,64 @@ impl IngredientIndices {
         &self,
         mut memo_index: impl FnMut(IngredientIndex) -> MemoIngredientIndex,
     ) -> MemoIngredientIndices {
-        let mut memo_ingredient_indices = [(
-            IngredientIndex::from((u32::MAX - 1) as usize),
+        let mut indices = Vec::new();
+        indices.resize(
+            self.indices.last().unwrap().as_usize() + 1,
             MemoIngredientIndex::from_usize((u32::MAX - 1) as usize),
-        ); MAX_MEMO_INGREDIENT_INDICES];
-        for i in 0..usize::from(self.len) {
-            let memo_ingredient_index = memo_index(self.indices[i]);
-            memo_ingredient_indices[i] = (self.indices[i], memo_ingredient_index);
+        );
+        for &ingredient in &self.indices {
+            indices[ingredient.as_usize()] = memo_index(ingredient);
         }
         MemoIngredientIndices {
-            indices: memo_ingredient_indices,
-            len: self.len,
+            indices: indices.into_boxed_slice(),
         }
     }
 
     #[inline]
-    pub fn uninitialized() -> Self {
+    pub fn empty() -> Self {
         Self {
-            indices: [IngredientIndex::from((u32::MAX - 1) as usize); MAX_MEMO_INGREDIENT_INDICES],
-            len: 0,
+            indices: Box::default(),
         }
     }
 
-    #[track_caller]
-    #[inline]
-    pub fn merge(&mut self, other: &Self) {
-        if usize::from(self.len) + usize::from(other.len) > MAX_MEMO_INGREDIENT_INDICES {
-            panic!("too many variants in the salsa enum");
+    pub fn merge(iter: impl IntoIterator<Item = Self>) -> Self {
+        let mut indices = Vec::new();
+        for index in iter {
+            indices.extend(index.indices);
+        }
+        indices.sort_unstable();
+        indices.dedup();
+        Self {
+            indices: indices.into_boxed_slice(),
         }
-        self.indices[usize::from(self.len)..][..usize::from(other.len)]
-            .copy_from_slice(&other.indices[..usize::from(other.len)]);
-        self.len += other.len;
     }
 }
 
 /// This type is to [`MemoIngredientIndex`] what [`IngredientIndices`] is to [`IngredientIndex`]:
 /// since enums can contain different ingredient indices, they can also have different memo indices,
 /// so we need to keep track of them.
+///
+/// This effectively acts a map from [`IngredientIndex`] to [`MemoIngredientIndex`] but implemented
+/// via a slice for fast lookups, this means we trade memory for speed, turning an O(n) lookup into
+/// into O(1).
```

Comment:
```suggestion
/// This acts a map from [`IngredientIndex`] to [`MemoIngredientIndex`] but implemented
/// via a slice for fast lookups, trading memory for speed. With these changes, lookups are `O(1)`
/// instead of `O(n)`.
```